File: 总题库/46.Permutations.py
# 回溯法
class Solution:
    def __init__(self):
        super().__init__()
        self.res = []

    def backtrack(self, nums: List[int], index: int) -> None:
        if index == len(nums) - 1:
            self.res.append(nums[::])

        for i in range(index, len(nums)):
            nums[index], nums[i] = nums[i], nums[index]
            self.backtrack(nums, index + 1)
            nums[index], nums[i] = nums[i], nums[index]

    def permute(self, nums: List[int]) -> List[List[int]]:
        self.backtrack(nums, 0)
        return self.res

File: 总题库/46.Permutations.cpp
// 回溯法
class Solution {
    vector<vector<int>> ans;

    void backtrack(vector<int>& nums, int index) {
        if (index == nums.size() - 1) {
            ans.emplace_back(nums);
        }

        for (int i = index; i < nums.size(); i++) {
            swap(nums[index], nums[i]);
            backtrack(nums, index + 1);
            swap(nums[index], nums[i]);
        }
    }

public:
    vector<vector<int>> permute(vector<int>& nums) {
        backtrack(nums, 0);
        return ans;
    }
};